### ISSUE TITLE
[Feature Request] xApplicationHostConfig to cover general purpose applicationhost.config configuration

### ISSUE THREAD

**Issue by guitarrapc** [PowerShell/xWebAdministration#170]
Hi, appreciate for this super cool DSC resources. It greatly save my time.

BTW, current resources are very nice with specific usages but it doesn't cover other settings like "advancedLogging", "applicationRequestRouting", "proxy", "fileExtensions", "requestFiltering" and others. I believe specific usage works very nice and improve development efficiency, however if I need to create Script Resource (or brand new resource) for those not covered with current resources.

These gap is quite big for applicationhost.config because applicationhost.config (means IIS) is extendable with modules. But I believe everyone don't want to write such a Script Resource for each.

I guess it's nice if more general purpose resource to configure applicationhost.config is exists. Here's some sample Script Resource I need write for your reference and people who wonder how to write with.
### system.applicationHost/sites/siteDefaults/advancedLogging

This is typical current resource design limitation. Yes there are [xWebSiteDefaults](https://github.com/PowerShell/xWebAdministration/blob/64df110da311c84cf40fb21e3862c3d4b944ca3c/DSCResources/MSFT_xWebSiteDefaults/MSFT_xWebSiteDefaults.schema.mof) to cover siteDefaults. But it doesn't support for module specific configurations because Advanced Logging is additional module, even this module is using for most of IIS Production Services.

Here's some sample resource just to change log location.

``` xml
<system.applicationHost>
    <sites>
        <siteDefaults>
            <advancedLogging directory="C:\SOME\WHERE\I\WANT\TO\KEEP\AdvancedLogs" />
        </siteDefaults>
    </sites>
</system.applicationHost>
```

``` powershell
configuration Hoge
{
    $directory = "C:\SOME\WHERE\I\WANT\TO\KEEP\AdvancedLogs";
    Script AdbancedLogingLogDirectory
    {
        SetScript = {
            Start-WebCommitDelay 
            Set-WebConfigurationProperty -Filter /system.applicationHost/sites/siteDefaults/advancedLogging -PSPath machine/webroot/apphost -Name directory -Value $using:directory
            Stop-WebCommitDelay -Commit $true 
        }
        TestScript = {
            return (Get-WebConfigurationProperty -Filter /system.applicationHost/sites/siteDefaults/advancedLogging -PSPath machine/webroot/apphost -Name directory).Value -eq $using:directory
        }
        GetScript = {
            @{
                GetScript = (Get-WebConfigurationProperty -Filter /system.applicationHost/sites/siteDefaults/advancedLogging -PSPath machine/webroot/apphost -Name directory).Value
                Result = (Get-WebConfigurationProperty -Filter /system.applicationHost/sites/siteDefaults/advancedLogging -PSPath machine/webroot/apphost -Name directory).Value -eq $using:directory
            }
        }
    }
}
```
### system.applicationHost/advancedLogging

Almost same as above but **advancedLogging** section is not supported with current resources.

``` xml
<configuration>
    <system.applicationHost>
        <advancedLogging>
            <serverLogs directory="C:\SOME\WHERE\I\WANT\TO\KEEP\AdvancedLogs" />
        </advancedLogging>
    </system.applicationHost>
</configuration>
```

``` powershell
configuration Hoge
{
    $directory = "C:\SOME\WHERE\I\WANT\TO\KEEP\AdvancedLogs";
    Script AdbancedLogingLogDirectory
    {
        SetScript = {
            Start-WebCommitDelay 
            Set-WebConfigurationProperty -Filter /system.applicationHost/advancedLogging/serverLogs  -PSPath machine/webroot/apphost -Name directory -Value $using:directory
            Stop-WebCommitDelay -Commit $true 
        }
        TestScript = {
            return (Get-WebConfigurationProperty -Filter /system.applicationHost/advancedLogging/serverLogs -PSPath machine/webroot/apphost -Name directory).Value -eq $using:directory
        }
        GetScript = {
            @{
                GetScript = (Get-WebConfigurationProperty -Filter /system.applicationHost/advancedLogging/serverLogs -PSPath machine/webroot/apphost -Name directory).Value
                Result = (Get-WebConfigurationProperty -Filter /system.applicationHost/advancedLogging/serverLogs -PSPath machine/webroot/apphost -Name directory).Value -eq $using:directory
            }
        }
    }
}
```
### location support

Current resources doesn't support **Location**. This let people write Script resource.

``` xml
<configuration>
    <location path="SITENAME">
        <system.webServer>
            <httpLogging dontLog="true" />
        </system.webServer>
    </location>
</configuration>
```

``` powershell
configuration Hoge
{
    $dontLog = $true
    $siteName = "SITENAME"
    Script HttpLogging
    {
        SetScript = {
            Start-WebCommitDelay 
            Set-WebConfigurationProperty -Filter /system.webserver/httpLogging -PSPath machine/webroot/apphost -Location $using:siteName  -Name dontLog -Value $using:dontLog
            Stop-WebCommitDelay -Commit $true 
        }
        TestScript = {
            return  (Get-WebConfigurationProperty -Filter /system.webserver/httpLogging -PSPath machine/webroot/apphost -Location $using:siteName  -Name dontLog).Value -eq $using:dontLog
        }
        GetScript = {
            @{
                GetScript = (Get-WebConfigurationProperty -Filter /system.webserver/httpLogging -PSPath machine/webroot/apphost -Location $using:siteName  -Name dontLog).Value
                Result = (Get-WebConfigurationProperty -Filter /system.webserver/httpLogging -PSPath machine/webroot/apphost -Location $using:siteName -Name dontLog).Value -eq $using:dontLog
            }
        }
    }
}
```
### system.webServer/rewrite

ARR is almost neccesary module for IIS to manage proxy or web farms. However script resource is required if anyone want to manage **rewrite**,  **requestRouting** and others. Also this resource is too much tricky, I don't think everyone can write this without knowledge for `Set-WebConfigurationProperty`.

``` xml
<configuration>
    <location path="SITENAME">
        <system.webServer>
            <rewrite>
                <allowedServerVariables>
                    <add name="VARIABLETOCREATE" />
                </allowedServerVariables>
            </rewrite>
        </system.webServer>
    </location>
</configuration>
```

``` powershell
configuration Hoge
{
    $allowedServerVariables = "VARIABLETOCREATE"
    $siteName = "SITENAME"
    Script ArrVariables
    {
        SetScript = {
            Start-WebCommitDelay 
            Set-WebConfigurationProperty -Filter /system.webserver/rewrite/allowedServerVariables -Location $using:siteName -Name "." -Value @{name = $using:allowedServerVariables}
            Stop-WebCommitDelay -Commit $true 
        }
        TestScript = {
            return  (Get-WebConfigurationProperty -Filter /system.webserver/rewrite/allowedServerVariables -Location $using:siteName -Name "." -Recurse).Collection.name -eq $using:allowedServerVariables
        }
        GetScript = {
            @{
                GetScript = (Get-WebConfigurationProperty -Filter /system.webserver/rewrite/allowedServerVariables -Location $using:siteName -Name "." -Recurse).Collection.name
                Result = (Get-WebConfigurationProperty -Filter /system.webserver/rewrite/allowedServerVariables -Location $using:siteName -Name "." -Recurse).Collection.name -eq $using:allowedServerVariables
            }
        }
    }
}
```

I believe core of every script resource is almost same and can be refactor to **General Purpose xApplicationHostConfig** resource. What do you think?

**Comment by johlju** [comment]
@guitarrapc Thanks for the detailed issue description, with this information it is hopefully simpler for someone to run with this. It would be good to get a proposed schema.mof for the proposed general purpose xApplicationHostConfig resource. Would this replace any existing resource?

**Comment by guitarrapc** [comment]
Yes, I do believe so.
mof would require Schema's several infos.

* filter
* location
* name
* value

Also Operator would be required.

* Ensure (Present = SET, Absent = REMOVE)

so the mof schema will be... similar to the following.

string Ensure : Key
string Filter : Required
string Location : Required
string Name : Required
string Value : Write

**Comment by regedit32** [comment]
This appears to now be resolved by PR #330.  Please re-open if you feel this needs to be further addressed.
